### ISSUE TITLE
Remote-SSH is stuck in an infinte nslookup loop

### ISSUE THREAD

**Issue by grawity** [microsoft/vscode#207480]
Does this issue occur when all extensions are disabled?: No

- VS Code Version: 1.87.1 
- OS Version: Windows 11

I'm aware of "Issues caused by an extension need to be reported directly to the extension publisher", but... Microsoft is the extension publisher in this case.

Steps to Reproduce:

1. Have an `~\.ssh\config` with these contents:

    ```
    Host foohostname
        # At least one host section needed; can remain empty.
    
    Host *
        # Canonicalization needs to be enabled.
        CanonicalizeHostname yes
        # More than one domain needs to be specified to trigger the issue.
        CanonicalDomains example.net example.com
    ```

2. Invoke "Remote-SSH: Connect to host..." or "Remote-SSH: Connect current window to host..."

4. The extension never activates. Tools such as ProcMon or ProcExp show that it keeps spawning `nslookup *.example.net` in a loop, the exact same command line repeating infinitely until I close VSCode.

(Upon inspecting extension.js, it seems like the extension looks for output `/server can't find/`, but nslookup doesn't output that exact text here – it outputs `HomeGW.home.arpa can't find` instead. Not sure whether that's related to the infinite loop.)

This is a new issue; it happens with Remote-SSH `v0.109.0` and with `v0.110.2024030815 (pre-release)`, but it used to work 2-3 months ago.

**Comment by roblourens** [comment]
Not sure what you're referring to re: `server can't find`. Can you just share the log from the Remote-SSH output channel?

**Comment by grawity** [comment]
> Can you just share the log from the Remote-SSH output channel?

It seems to only contain:

```
[15:40:33.796] Log Level: 2
[15:40:33.810] VS Code version: 1.87.2
[15:40:33.810] Remote-SSH version: remote-ssh@0.109.0
[15:40:33.810] win32 x64
[15:40:33.812] Picking SSH host
```

`code --log trace` doesn't add anything more.

(To clarify, it doesn't fail to _connect_ – it fails to even _ask for host name,_ i.e. after I select "Remote-SSH: Connect to Host...", the hostname entry field doesn't pop up.)

**Comment by roblourens** [comment]
Not sure what the problem is, can you tell me more about 

> Tools such as ProcMon or ProcExp show that it keeps spawning nslookup *.example.net in a loop, the exact same command line repeating infinitely until I close VSCode.

I'm pretty sure that the extension is not doing this. Is it possible that there is something else on your machine, eg antivirus, that could be interfering?

**Comment by grawity** [comment]
> I'm pretty sure that the extension is not doing this

Well, the reason I mentioned that is... the file `C:\Users\grawity\.vscode\extensions\ms-vscode-remote.remote-ssh-0.109.0\out\extension.js` literally has a call to nslookup in it:

```javascript
... {stdout:s}=(0,i.spawnSync)("nslookup",[o]);if(!/server can't find/.test((s.toString())) ...
```

I also have a Linux machine with VSCode, and I can see that it _also_ runs a `nslookup *.whatever` when I select "Remote-SSH: Connect to Host..." (although it doesn't get into an infinite loop there, so the host list popup opens correctly after a slight delay).

> Is it possible that there is something else on your machine, eg antivirus, that could be interfering?

Unlikely; the only antivirus software I have is the built-in Windows Defender, and I also did (for unrelated reasons) completely reinstall Win11 in between the last two messages, so currently there's very little unusual software on the system.

**Comment by roblourens** [comment]
Oh, I didn't realize that the ssh-config library does this when you enable canonicalization. I can't repro an issue when the domain is found or not found, but I'm guessing it's related to your different output, `HomeGW.home.arpa can't find`.

**Comment by roblourens** [comment]
I think I could see how it could possibly loop, if it gets canonicalized to a value that still matches the 'Host' block with the canonicalization rule, and tries to re-canonicalize forever. Are those the actual values in your config file, or can you share the actual values?

**Comment by grawity** [comment]
> Are those the actual values in your config file, or can you share the actual values? I can file an issue upstream if I can reproduce it.

Yes and no, it's not the original config but it's still an actual config that I used to reproduce the issue before posting. I'm pretty sure any 2+ domains work (could not reproduce with only 1 domain).

The "original" ssh/config at home [is here](https://gist.github.com/grawity/2c989214f6b81130a31fe42912ca0c5c), but I tried again on my work PC this time, and was still able to reproduce it with literally `CanonicalDomains example.com example.net` plus a dummy `Host asdadafdg` section.

I *think* the only external requirement is that Windows must be using a DNS server whose IP address has "reverse DNS" available (like how my home router has 192.168.1.1 resolve to 'HomeGW.home.arpa' and Google's 8.8.8.8 resolves to 'dns.google') so that instead of `** server cannot find blah` it would show `** dns.google cannot find blah`.

(Linux is immune because its nslookup clone is hardcoded to say 'server', it seems.)

**Comment by roblourens** [comment]
I couldn't repro with Windows + Google DNS, but I filed an issue upstream: https://github.com/cyjake/ssh-config/issues/81
